tracee/tracee.bpf.c

```diff
@@ -498,26 +500,40 @@ static __always_inline int get_config(u32 key)
     return *config;
 }
 
-// returns 1 if you should trace based on uid
+// returns 1 if you should trace based on uid, 0 if not
 static __always_inline int uid_filter_matches()
 {
+
+    int equal;
+    int not_equal;
```

Comment:
Do we really need to have both of these vars?
Seems redundant to me

Reply:
Thinking about how I can make it more elegant, but the reason is because `NotEqual` does not mean `Equal`  was not specified

Reply:
Right, but in this case, you have a given uid, which only has one value in uid_equal_filter map.
If equal=1, then not_equal=0 and the other way around, right?
If so - there is no new information in the second variable.
Actually, I think non of these variables are needed, and let me explain.
Thinking about the formula you came up with:  (=) || (> && < && !(!=))
If uid value in uid_equal_filter is not NULL, we can return right away with the correct result.

Reply:
So I find it confusing, but the issue with this is that if a uid doesn't have any matching filters, the event will be allowed. But in the case of lets say:

User specifies: `--filter uid=0` (user only wants events with uid=0)

If lets say an event has a uid 1000 it would pass ('equality', 'lessThan', 'greaterThan' all NULL). If we return `0` at the end it breaks the comparison filters.

Let me see how I can clean it up...

Reply:
Good point, but I think it has an easy fix.
To avoid missing cases, I tried to rewrite this function:
static __always_inline int uid_filter_matches()
{
    uid_t uid = (u32)bpf_get_current_uid_gid();
    s64* equality = bpf_map_lookup_elem(&uid_equal_filter, &uid);
    if (equality != NULL) {
        return *equality;
    }

    u32 greater = GREATER;
    u32 less = LESS;

    s64* greaterThan = bpf_map_lookup_elem(&uid_compare_filter, &greater);
    s64* lessThan = bpf_map_lookup_elem(&uid_compare_filter, &less);

    if (lessThan != NULL) {
        if ((uid >= *lessThan)) 
            return 0;
    }

    if (greaterThan != NULL ) { 
        if ((uid <= *greaterThan)) 
            return 0;
    }

    if (greaterThan == NULL && lessThan == NULL)
        // If we reached this point, there is no filter match
        return 0;

    return 1;
}

If may have missed something, so please tell me what you think

Reply:
It looks good to me, but I'm being thrown for a serious loop by the verifier. When running your code I get:

```
257: (15) if r3 == 0x0 goto pc+3
 R0=map_value(id=0,off=0,ks=4,vs=8,imm=0) R1_w=inv(id=0) R2_w=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R3_w=map_value(id=0,off=0,ks=4,vs=8,imm=0) R6=invP0 R7=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R8=inv0 R9=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R10=fp0 fp-80=????mmmm fp-88=????mmmm fp-104=mmmmmmmm fp-112=mmmmmmmm fp-120=mmmmmmmm fp-128=mmmmmmmm fp-136=mmmmmmmm fp-144=mmmmmmmm fp-152=mmmmmmmm fp-160=mmmmmmmm fp-168=mmmmmmmm fp-176=mmmmmmmm fp-184=mmmmmmmm fp-192=mmmmmmmm fp-200=mmmmmmmm fp-208=mmmmmmmm fp-216=mmmmmmmm fp-224=mmmmmmmm fp-232=mmmmmmmm fp-240=mmmmmmmm fp-248=mmmmmmmm fp-256=mmmmmmmm fp-264=mmmmmmmm fp-272=mmmm???? fp-368=????mmmm fp-376=ctx fp-384=mmmmmmmm fp-392=map_value
; if ((greaterThan != NULL) && (uid <= *greaterThan)) { 
258: (79) r1 = *(u64 *)(r3 +0)
 R0=map_value(id=0,off=0,ks=4,vs=8,imm=0) R1_w=inv(id=0) R2_w=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R3_w=map_value(id=0,off=0,ks=4,vs=8,imm=0) R6=invP0 R7=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R8=inv0 R9=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R10=fp0 fp-80=????mmmm fp-88=????mmmm fp-104=mmmmmmmm fp-112=mmmmmmmm fp-120=mmmmmmmm fp-128=mmmmmmmm fp-136=mmmmmmmm fp-144=mmmmmmmm fp-152=mmmmmmmm fp-160=mmmmmmmm fp-168=mmmmmmmm fp-176=mmmmmmmm fp-184=mmmmmmmm fp-192=mmmmmmmm fp-200=mmmmmmmm fp-208=mmmmmmmm fp-216=mmmmmmmm fp-224=mmmmmmmm fp-232=mmmmmmmm fp-240=mmmmmmmm fp-248=mmmmmmmm fp-256=mmmmmmmm fp-264=mmmmmmmm fp-272=mmmm???? fp-368=????mmmm fp-376=ctx fp-384=mmmmmmmm fp-392=map_value
; if ((greaterThan != NULL) && (uid <= *greaterThan)) { 
259: (61) r2 = *(u32 *)(r10 -368)
; if ((greaterThan != NULL) && (uid <= *greaterThan)) { 
260: (7d) if r1 s>= r2 goto pc-64
 R0=map_value(id=0,off=0,ks=4,vs=8,imm=0) R1=inv(id=0) R2=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R3=map_value(id=0,off=0,ks=4,vs=8,imm=0) R6=invP0 R7=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R8=inv0 R9=inv(id=0,umax_value=4294967295,var_off=(0x0; 0xffffffff)) R10=fp0 fp-80=????mmmm fp-88=????mmmm fp-104=mmmmmmmm fp-112=mmmmmmmm fp-120=mmmmmmmm fp-128=mmmmmmmm fp-136=mmmmmmmm fp-144=mmmmmmmm fp-152=mmmmmmmm fp-160=mmmmmmmm fp-168=mmmmmmmm fp-176=mmmmmmmm fp-184=mmmmmmmm fp-192=mmmmmmmm fp-200=mmmmmmmm fp-208=mmmmmmmm fp-216=mmmmmmmm fp-224=mmmmmmmm fp-232=mmmmmmmm fp-240=mmmmmmmm fp-248=mmmmmmmm fp-256=mmmmmmmm fp-264=mmmmmmmm fp-272=mmmm???? fp-368=????mmmm fp-376=ctx fp-384=mmmmmmmm fp-392=map_value
; if ((greaterThan != NULL) && (uid <= *greaterThan)) { 
261: (4f) r3 |= r0
R3 pointer |= pointer prohibited
processed 966 insns (limit 1000000) max_states_per_insn 2 total_states 67 peak_states 67 mark_read 26
```

I suspected this has to do with checking `greaterThan` and `lessThan` after dereferencing the pointers, but I think I disproved that. My other theory was the third check (`(greaterThan == NULL && lessThan == NULL)`) quick fails if `greaterThan` is NULL so technically lessThan isn't checked but they're checked above regardless.

Is this happening when you run on your machine? I'm on mainline `5.9.11-200` kernel.

Reply:
Just checked on my machine, and I don't get verifier problems.
Can you please try to break the if:

```
    if ((lessThan != NULL) {
        if (uid >= *lessThan))
            return 0;
    }

    if ((greaterThan != NULL) {
        if (uid <= *greaterThan))
            return 0;
    }
```

Reply:
First thing I tried, same issue...

Reply:
What's strange is that if I remove the third if statement I don't get the verifier issue, but if I add it back the error message has to do with the first if statement.

Reply:
Which clang version do you use?
It seems to be related to the registers assignment (I think)

Reply:
I have 11.0.0 installed

Reply:
I now also see a logical problem here.
If no uid filter was chosen, greaterThan and lessThan will be null, and will get no output event

Reply:
Yea I noticed that too. Locally I changed the Go code to have the base uid filter start open and to always Update the map in tracee.go.

Reply:
Can you please push your latest changes so I can test them locally?

Reply:
Done